File: src/main/java/com/tibiawiki/domain/objects/WikiObject.java
package com.tibiawiki.domain.objects;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.tibiawiki.domain.enums.Article;
import com.tibiawiki.domain.enums.Status;
import com.tibiawiki.domain.interfaces.Validatable;
import com.tibiawiki.domain.objects.validation.ValidationResult;
import lombok.Getter;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@Getter
public abstract class WikiObject implements Validatable {

    private final String templateType;
    private final String name;
    private final Article article;
    private final String actualname;
    private final String plural;
    private final String implemented;
    private final String notes;
    private final String history;
    private final Status status;

    protected WikiObject() {
        templateType = null;
        name = null;
        article = null;
        actualname = null;
        plural = null;
        implemented = null;
        notes = null;
        history = null;
        status = null;
    }

    @SuppressWarnings("squid:S00107")
    public WikiObject(String name, Article article, String actualname, String plural, String implemented, String notes,
                      String history, Status status) {
        this.templateType = null;
        this.name = name;
        this.article = article;
        this.actualname = actualname;
        this.plural = plural;
        this.implemented = implemented;
        this.notes = notes;
        this.history = history;
        this.status = status;
    }

    public abstract List<String> fieldOrder();

    @JsonIgnore
    public String getClassName() {
        return this.getClass().getSimpleName();
    }

    @Override
    public String toString() {
        return "Class: " + getClassName() + ", name: " + getName();
    }

    @Override
    public List<ValidationResult> validate() {
        return Collections.emptyList();
    }

    public static class WikiObjectImpl extends WikiObject {

        public WikiObjectImpl() {
            super();
        }

        @Override
        public List<String> fieldOrder() {
            return Arrays.asList("name", "article", "actualname", "plural", "implemented", "notes", "history", "status");
        }

        @Override
        public List<ValidationResult> validate() {
            return Collections.emptyList();
        }
    }
}

File: src/main/kotlin/com/tibiawiki/domain/objects/WikiObject.kt
package com.tibiawiki.domain.objects

import com.fasterxml.jackson.annotation.JsonIgnore
import com.tibiawiki.domain.enums.Article
import com.tibiawiki.domain.enums.InfoboxTemplate
import com.tibiawiki.domain.enums.Status
import com.tibiawiki.domain.interfaces.Validatable
import com.tibiawiki.domain.objects.validation.ValidationResult
import com.tibiawiki.domain.objects.validation.ValidationSeverity

abstract class WikiObject(
    open val name: String? = null,
    open val article: Article? = null,
    open val actualname: String? = null,
    open val plural: String? = null,
    open val implemented: String? = null,
    open val notes: String? = null,
    open val history: String? = null,
    open val status: Status? = null
) : Validatable {

    abstract fun fieldOrder(): List<String>

    abstract fun getTemplateType(): String

    @get:JsonIgnore
    val className: String
        get() = this::class.simpleName.orEmpty()

    override fun toString(): String {
        return "Class: $className, name: $name"
    }

    /**
     * Wiki page title used for reads/writes. Defaults to [name]; types that key
     * off another field (e.g. Key number) override this.
     */
    @JsonIgnore
    open fun articleTitle(): String {
        return name?.trim().orEmpty()
    }

    override fun validate(): List<ValidationResult> {
        val results = mutableListOf<ValidationResult>()
        if (articleTitle().isBlank()) {
            results += ValidationResult(ValidationSeverity.ERROR, NAME_REQUIRED)
        }
        val templateType = getTemplateType()
        if (templateType !in KNOWN_TEMPLATE_TYPES) {
            results += ValidationResult(ValidationSeverity.ERROR, "$UNKNOWN_TEMPLATE_TYPE$templateType")
        }
        return results
    }

    class WikiObjectImpl : WikiObject() {
        override fun fieldOrder(): List<String> {
            return listOf("name", "article", "actualname", "plural", "implemented", "notes", "history", "status")
        }

        override fun getTemplateType(): String {
            return "WikiObjectImpl"
        }
    }

    companion object {
        const val NAME_REQUIRED = "name is required"
        const val UNKNOWN_TEMPLATE_TYPE = "unknown templateType: "
        val KNOWN_TEMPLATE_TYPES: Set<String> = InfoboxTemplate.entries.map { it.templateName }.toSet()
    }
}
